### ISSUE TITLE
[3.1] FormHelper & Nesting Validators: No error messages in the form

### ISSUE THREAD

**Issue by OopsDoh** [cakephp/cakephp#7532]
When using [nested validators](http://book.cakephp.org/3.0/en/core-libraries/validation.html#nesting-validators) the error messages for the nested fields do exist, but they are not shown in the form. I think the form helper [Display and Checking Errors](http://book.cakephp.org/3.0/en/views/helpers/form.html#displaying-and-checking-errors) can't handle the "nested" error messages and field names like `Model.nested.array.fieldname`.

Example: I'm trying to apply validation rules for translation input fields. Let's assume the regular `title` field has `minLength` of 5 and the translated field should have a `minLength` of 10.

```php
// The nested validation is set up in the table class as described in the book.
// http://book.cakephp.org/3.0/en/core-libraries/validation.html#nesting-validators

// Example input data array ($this->request->data)
[
  'title' => 'abc',
  '_translations' => [
    'fr' => [
      'title' => 'abc'
    ],
  ]
]

// Then in controller
...
$item = $this->Items->patchEntity($item, $this->request->data);
debug($item);
...

// The validation is working, errors are available in the debug result:
object(App\Model\Entity\Item) {
	'[new]' => true,
	'[accessible]' => [
		'*' => true
	],
	'[dirty]' => [],
	'[original]' => [],
	'[virtual]' => [],
	'[errors]' => [
		'title' => [
			'minLength' => 'minLengh is set to 5'
		],
		'_translations' => [
			'fr' => [
				'title' => [
					'minLength' => 'minLengh is set to 10.'
				]
			]
		],
	],
	'[repository]' => 'Items'
}

// But in the form (view) you can't show the error:
echo $this->Form->error('items._translations.fr.title');
// Result: Nothing :(
// Also the input field does not get marked with the css error class...
```

**Comment by markstory** [comment]
I took a look into this and its a bit more complex than I had originally anticipated. The `isRequired()` method also needs some rethinking to handle nested validators. Additionally the FormContext needs similar rework.

**Comment by beporter** [comment]
The fix for `FormContext` might be pretty straightforward. I just "solved" it in my current project with this diff:

```diff
diff --git a/src/View/Form/FormContext.php b/src/View/Form/FormContext.php
index c4d3893..77c4513 100644
--- a/src/View/Form/FormContext.php
+++ b/src/View/Form/FormContext.php
@@ -135,8 +135,9 @@ class FormContext implements ContextInterface
     public function error($field)
     {
         $errors = $this->_form->errors();
-        if (isset($errors[$field])) {
-            return array_values($errors[$field]);
+        $errors = \Cake\Utility\Hash::extract($errors, $field);
+        if (!empty($errors)) {
+            return $errors;
         }
         return [];
     }
```

I have not run the test suite, but if it works without side effects I can try to get a PR done sometime this week.

**Comment by markstory** [comment]
@beporter Awesome. I started off with the EntityContext and it got out of hand really quickly.

**Comment by lorenzo** [comment]
@beporter Would you mind submitting that patch for the form context?

**Comment by beporter** [comment]
PR submitted, currently waiting on the builds.

**Comment by ADmad** [comment]
Closing as related PR is open.

**Comment by beporter** [comment]
Just for the record, the PR doesn't fully address the issues mentioned here.

**Comment by OopsDoh** [comment]
Thank you @beporter for the form context fix! With this fix the error messages for nested fields should be visible in the form, did I get this right? Unfortunately I'm not sure if it's working, at least not in my test.

I just made a new local installation of CakePHP 3.1.6 to test it. First with translate behavior, but the errors are **still not visible**, so I also added an additional nested test input field (unrelated to the translate behaviour), unfortunately also no error message output... 

Debug array with error messages in controller (looking ok, all fields are validated separately):
```
object(App\Model\Entity\Article) {
	'[new]' => true,
	'[accessible]' => [
		'*' => true
	],
	'[dirty]' => [],
	'[original]' => [],
	'[virtual]' => [],
	'[errors]' => [
		'title' => [
			'length' => 'English Title (Default) need to be at least 5 characters long.'
		],
		'_translations' => [
			'es' => [
				'title' => [
					'length' => 'Spanish Title need to be at least 10 characters long.'
				]
			],
			'de' => [
				'title' => [
					'length' => 'German Title need to be at least 15 characters long.'
				]
			]
		],
		'simple' => [
			'test' => [
				'length' => 'Testfield need to be at least 20 characters long.'
			]
		]
	],
	'[repository]' => 'Articles'
}
```
But in the view there is only the error message for the **not-nested** field available...

```php
echo '<p>Error messages test output:<br />';
echo 'English: '.$this->Form->error('title').'</br>'; // CORRECT error output
echo 'Spanish: '.$this->Form->error('_translations.es.title').'<br />'; // No error visible
echo 'German: '.$this->Form->error('_translations.de.title').'<br />'; // No error visible
echo 'Test Field: '.$this->Form->error('simple.test').'</p>'; // No error visible
```
Here is my test setup (I simplified my code for GitHub, hope I didn't remove an important part):

**src/Template/Articles/add.ctp**:
```php
echo $this->Form->create($article);
 // Nested, but not a real field in the database
echo $this->Form->input('simple.test', ['label' => 'Test field']);
// In Articles table:
echo $this->Form->input('title', ['label' => 'Title English']);
// i18n translate behaviour fields:
echo $this->Form->input('_translations.es.title', ['label' => 'Title Spanish']);
echo $this->Form->input('_translations.de.title', ['label' => 'Title German']);
echo $this->Form->button(__('Submit'));
echo $this->Form->end();
```

**src/Controller/ArticlesController.php**
```php
public function add() {
  $article = $this->Articles->newEntity();
  if ($this->request->is('post')) {
    debug($this->request->data); // Data array is looking ok
    $article = $this->Articles->patchEntity($article, $this->request->data);
    debug($article);  // Validation errors are all available
    if ($this->Articles->save($article)) {
      $this->Flash->success(__('The article has been saved.'));
    } else {
      $this->Flash->error(__('The article could not be saved. Please, try again.'));
    }
  }
  $this->set(compact('article'));
  $this->set('_serialize', ['article']);
}
```

**src/Model/Table/ArticlesTable.php**
```php
public function validationDefault(Validator $validator) {
  $validator
    ->requirePresence('title', 'create')
    ->notEmpty('title')
    ->add('title', [
      'length' => [
        'rule' => ['minLength', 5],
        'message' => __('English Title (Default) need to be at least 5 characters long.')
      ]
    ]);

  // Nested Validation TEST FIELD
  $testValidator = new Validator();
  $testValidator
    ->add('test', [
      'length' => [
      'rule' => ['minLength', 20],
      'message' => __('Testfield need to be at least 20 characters long.')
      ]]);
  $validator
    ->addNested('simple', $testValidator);

  // Nested Validation for translate behaviour fields
  $translationValidator = new Validator();
  // Spanish title field
  $spanishValidator = new Validator();
  $spanishValidator
    ->add('title', [
      'length' => [
        'rule' => ['minLength', 10],
        'message' => __('Spanish Title need to be at least 10 characters long.')
      ]]);
  // German title field
  $germanValidator = new Validator();
  $germanValidator
    ->add('title', [
      'length' => [
        'rule' => ['minLength', 15],
        'message' => __('German Title need to be at least 15 characters long.')
      ]]);
  // Combine validators
  $translationValidator
    ->addNested('es', $spanishValidator)
    ->addNested('de', $germanValidator);
  $validator
    ->addNested('_translations', $translationValidator);
  // Done
  return $validator;
}
```

**Comment by markstory** [comment]
@OopsDoh The `FormContext` object is used when making forms for model-less forms. ORM backed forms are still not working.

**Comment by robertscherer** [comment]
Here's my dirty, but simple solution for the problem - we're using JsonType fields a lot and by overriding FormHelper's `error()` method like this we get error messages:

    public function error($field, $text = null, array $options = [])
    {
        $parts = explode('.', $field);
        if (count($parts) > 1 && is_array($this->context()->entity()->get($parts[0]))) {
            $errors = $this->context()->entity()->errors();
            if (Hash::check($errors, $field)) {
                $error = Hash::extract($errors, $field);
                return $this->formatTemplate('error', ['content' => $error]);
            }
        }
        return parent::error($field, $text, $options);
    }

**Comment by markstory** [comment]
@robertscherer Interesting, I tried an approach similar to how we get values out of entities, but ran into a bunch of issues with the various association types.

**Comment by robertscherer** [comment]
@markstory It's the `is_array` check on the entity field value which should prevent these issues. OTOH, I didn't build any complex forms yet with this workaround yet.

**Comment by markstory** [comment]
As of 3.3.0 (#9225) TranslateBehavior can have a `validator` option set. This validator will be used to validate translated records, and the errors are propagated up to the root entity so `articles._translations.fr.title` will work as you'd expect.

**Comment by darensipes** [comment]
@markstory When you have an Entity (ORM backed) Form and that Entity/Table has a json column, How can you get the errors to show up on the form. Using nested validators all seems to work on the entity/table side but those errors never show up on the inputs. I have fixed before by just converting my json to php array and doing `new Entity($jsonData)`, for each nested level and copying the errors to the nested entities.

Kinda like:
```php
$article = $this->Articles->patchEntity($article, $this->request->data);
$entity = new Entity($this->request->data['meta']);
$entity->errors($article->errors('meta'));
$medicalUpload->set('meta', $entity);
```

It works, but I don't like my current solution. Is this something that could or should be fixed within CakePHP or is using a json column with nested levels out of scope with how the validation and EntityContext should work with the FormHelper. I don't mind testing or taking a stab at PR for this, but not if the consensus is it should be handled at the app level by overriding or customizing.

Like @robertscherer mentioned above, I am also using JsonType a lot in my application.

Thoughts?

**Comment by markstory** [comment]
> It works, but I don't like my current solution. Is this something that could or should be fixed within CakePHP or is using a json column with nested levels out of scope with how the validation and EntityContext should work with the FormHelper. 

I don't think we've ever made an explicit decision that they shouldn't work. I personally don't use JSON types so I'm don't have much input here. I can see how having validation messages from nested validators on a JSON column 'just work' would be quite helpful though.

If you can find a solution that doesn't introduce a significant amount of complexity I'd be open to merging it.

**Comment by Tartoob** [comment]
Hello,

I did my own solution

`	namespace Cake\View\Helper;

	class NestedFormHelper extends FormHelper
	{
		public function error($field, $text = null, array $options = [])
		{
			if (substr($field, -5) === '._ids') {
				$field = substr($field, 0, -5);
			}
			$options += ['escape' => true];

			$context = $this->_getContext();
			$parts = explode('.', $field);
			if (count($parts) > 1) {
				$errors = $this->context()->entity()->errors($parts[0]);
				if (!$errors || !isset($errors[$parts[1]]))
					return '';
					
				$error = (array)$errors[$parts[1]];
			} else {
				if (!$context->hasError($field)) {
					return '';
				}
				$error = (array)$context->error($field);
			}

			if (is_array($text)) {
				$tmp = [];
				foreach ($error as $k => $e) {
					if (isset($text[$k])) {
						$tmp[] = $text[$k];
					} elseif (isset($text[$e])) {
						$tmp[] = $text[$e];
					} else {
						$tmp[] = $e;
					}
				}
				$text = $tmp;
			}

			if ($text !== null) {
				$error = $text;
			}

			if ($options['escape']) {
				$error = h($error);
				unset($options['escape']);
			}

			if (is_array($error)) {
				if (count($error) > 1) {
					$errorText = [];
					foreach ($error as $err) {
						$errorText[] = $this->formatTemplate('errorItem', ['text' => $err]);
					}
					$error = $this->formatTemplate('errorList', [
						'content' => implode('', $errorText)
					]);
				} else {
					$error = array_pop($error);
				}
			}

			return $this->formatTemplate('error', ['content' => $error]);
		}
	}
`

And here is View form
`	<div class="customers form large-9 medium-8 columns content">
		<?= $this->NestedForm->create($customer) ?>
		<fieldset>
			<legend><?= __('Add Customer') ?></legend>
			<?php
				echo $this->NestedForm->control('user.name');
				echo $this->NestedForm->control('user.username');
				echo $this->NestedForm->control('user.password');
				echo $this->NestedForm->control('user.email');
				echo $this->NestedForm->control('title');
				echo $this->NestedForm->control('days');
			?>
		</fieldset>
		<?= $this->NestedForm->button(__('Submit')) ?>
		<?= $this->NestedForm->end() ?>
	</div>
`

**Comment by burzum** [comment]
@markstory can we re-open this and try @robertscherer solution? We have exactly the same use case here that we need to validate nested data.

**Comment by markstory** [comment]
@burzum Sure. I won't have time to look into that approach. But if you have the need/time go for it :smile:

**Comment by qqqqb** [comment]
@burzum have you tried my solution?

**Comment by burzum** [comment]
@qqqqb no because @robertscherer solution just works fine for me. Any reason why I should pick yours over his?

**Comment by qqqqb** [comment]
Sorry I thought you still have issue on this-------- Original Message --------Subject: Re: [cakephp/cakephp] [3.1] FormHelper & Nesting Validators: No error messages in the form (#7532)From: Florian Krämer To: cakephp/cakephp CC: F Alkaabi ,Mention @qqqqb no because @robertscherer solution just works fine for me. Any reason why I should pick yours over his?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/cakephp/cakephp","title":"cakephp/cakephp","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/cakephp/cakephp"}},"updates":{"snippets":[{"icon":"PERSON","message":"@burzum in #7532: @qqqqb no because @robertscherer solution just works fine for me. Any reason why I should pick yours over his?"}],"action":{"name":"View Issue","url":"https://github.com/cakephp/cakephp/issues/7532#issuecomment-347661896"}}}
